<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.11.0"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>NEFFy: NEFFy: NEFF Calculator and MSA File Converter</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<script type="text/javascript" src="clipboard.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="cookie.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  extensions: ["tex2jax.js"],
  jax: ["input/TeX","output/HTML-CSS"],
});
</script>
<script type="text/javascript" async="async" src="https://cdn.jsdelivr.net/npm/mathjax@2/MathJax.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<style>
  #titlearea {
    display: flex;
    justify-content: space-between;
    align-items: center;
    margin-top: 20px; /* Add top margin */
  }
  #projectinfo {
    display: flex;
    align-items: center;
  }
  #github-link {
    margin-left: auto;
    margin-right: 20px; /* Add a small right margin */
  }
  .green-text {
    color: green; /* Set text color to green */
  }
</style>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
  <div id="projectinfo">
    <div id="projectalign">
      <div id="projectname">NEFFy</div>
      <div id="projectbrief">NEFF Calculator and MSA File Converter</div>
    </div>
  </div>
  <div id="github-link">
    <a href="https://github.com/Maryam-Haghani/Neffy">
      <svg height="32" class="octicon octicon-mark-github" viewBox="0 0 16 16" version="1.1" width="32" aria-hidden="true">
        <path fill-rule="evenodd" d="M8 0C3.58 0 0 3.58 0 8c0 3.54 2.29 6.53 5.47 7.59.4.07.55-.17.55-.38 0-.19-.01-.82-.01-1.49-2.01.37-2.53-.49-2.69-.94-.09-.23-.48-.94-.82-1.13-.28-.15-.68-.52-.01-.53.63-.01 1.08.58 1.23.82.72 1.21 1.87.87 2.33.66.07-.52.28-.87.51-1.07-1.78-.2-3.64-.89-3.64-3.95 0-.87.31-1.59.82-2.15-.08-.2-.36-1.02.08-2.12 0 0 .67-.21 2.2.82.64-.18 1.32-.27 2-.27.68 0 1.36.09 2 .27 1.53-1.04 2.2-.82 2.2-.82.44 1.1.16 1.92.08 2.12.51.56.82 1.27.82 2.15 0 3.07-1.87 3.75-3.64 3.95.29.25.54.73.54 1.48 0 1.07-.01 1.93-.01 2.2 0 .21.15.46.55.38A8.013 8.013 0 0016 8c0-4.42-3.58-8-8-8z"></path>
      </svg>
      Edit on GitHub
    </a>
  </div>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.11.0 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search/",'.html');
/* @license-end */
</script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() { codefold.init(0); });
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search',false);
  $(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function(){ initResizable(false); });
/* @license-end */
</script>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<div id="MSearchResults">
<div class="SRPage">
<div id="SRIndex">
<div id="SRResults"></div>
<div class="SRStatus" id="Loading">Loading...</div>
<div class="SRStatus" id="Searching">Searching...</div>
<div class="SRStatus" id="NoMatches">No Matches</div>
</div>
</div>
</div>
</div>

</div><!-- top -->
<div id="doc-content">
<div><div class="header">
  <div class="headertitle"><div class="title">NEFFy: NEFF Calculator and MSA File Converter</div></div>
</div><!--header-->
<div class="contents">
<div class="textblock"><p><a class="anchor" id="autotoc_md0"></a>NEFFy is a versatile and efficient tool for bioinformatics research, offering advanced features for calculating NEFF (Normalized Effective Number of Sequences) for Multiple <a class="el" href="struct_sequence.html">Sequence</a> Alignments (MSA)s of any biological sequences, including protein, RNA, and DNA across various MSA formats. <br  />
 Additionally, NEFFy includes built-in support for format conversion, allowing users to seamlessly convert between different MSA formats.</p>
<h1><a class="anchor" id="autotoc_md1"></a>
Table of Contents</h1>
<ul>
<li><a class="el" href="index.html#installation">Installation</a></li>
<li>Tutorial<ul>
<li>Project Outline</li>
<li>Usage<ul>
<li>NEFF Computation</li>
<li>MSA File Conversion</li>
</ul>
</li>
</ul>
</li>
<li>Supported File Formats</li>
<li>Error Handling</li>
<li>License</li>
</ul>
<h1><a class="anchor" id="autotoc_md2"></a>
Installation</h1>
<p>To install the NEFFy tool, clone the repository and compile the code using a C++ compiler that supports C++17 or a newer version. You can use the provided [Makefile](Makefile) in the repository for this purpose. Navigate to the repository directory and enter the following command in the terminal:</p>
<div class="fragment"><div class="line">make</div>
</div><!-- fragment --><p> If the <code>make</code> command is not available on your operating system, <a href="https://maryam-haghani.github.io/Neffy/help.html#make_installation">here</a> is how you can install it.</p>
<p>Once the compilation is complete, you can run the program via the command line. <br  />
 This package is cross-platform and works on Linux, Windows, and macOS without requiring additional compilation. <br  />
</p>
<p>For more information on installing the executable, please refer to the <a href="https://maryam-haghani.github.io/Neffy/installation_guide.html#executable_install">documentation</a>.</p>
<h1><a class="anchor" id="autotoc_md3"></a>
Tutorial</h1>
<p>This tutorial will first explain the structure of the NEFFy repository, followed by a guide on how to use it.</p>
<h2><a class="anchor" id="autotoc_md4"></a>
Project Outline</h2>
<p>The NEFFy repository is structured as follows:</p>
<p><img src="docs/figs/outline.png" alt="outline" class="inline"/></p>
<h2><a class="anchor" id="autotoc_md5"></a>
Usage</h2>
<h2><a class="anchor" id="autotoc_md6"></a>
1. NEFF Computation</h2>
<p>NEFF determines the effective number of homologous sequences within a Multiple <a class="el" href="struct_sequence.html">Sequence</a> Alignment (MSA). It accounts for sequence similarities and provides a measure of sequence diversity. <br  />
 To calculate NEFF, use the <code>neff</code> program, provide an MSA file, and specify the desired flags.</p>
<h3><a class="anchor" id="autotoc_md7"></a>
Flags:</h3>
<p>The code accepts the following command-line flags: </p><table class="markdownTable">
<tr class="markdownTableHead">
<th class="markdownTableHeadNone">Flag   </th><th class="markdownTableHeadNone">Description   </th><th class="markdownTableHeadNone">Required   </th><th class="markdownTableHeadNone">Default Value   </th><th class="markdownTableHeadNone">Example    </th></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone"><code>--file=&lt;filename&gt;</code>   </td><td class="markdownTableBodyNone">Input file containing multiple sequence alignment   </td><td class="markdownTableBodyNone">Yes   </td><td class="markdownTableBodyNone">N/A   </td><td class="markdownTableBodyNone"><code>--file=my_alignment.fasta</code>    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyNone"><code>--alphabet=&lt;value&gt;</code>   </td><td class="markdownTableBodyNone">Alphabet of MSA <br  />
 <b>0</b>: Protein <br  />
 <b>1</b>: RNA <br  />
 <b>2</b>: DNA   </td><td class="markdownTableBodyNone">No   </td><td class="markdownTableBodyNone">0   </td><td class="markdownTableBodyNone"><code>--alphabet=1</code>    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone"><code>--check_validation=[true/false]</code>   </td><td class="markdownTableBodyNone">Validate the input MSA file based on alphabet or not   </td><td class="markdownTableBodyNone">No   </td><td class="markdownTableBodyNone">false   </td><td class="markdownTableBodyNone"><code>--check_validation=true</code>    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyNone"><code>--threshold=&lt;value&gt;</code>   </td><td class="markdownTableBodyNone">Threshold value of considering two sequences similar (between 0 and 1)   </td><td class="markdownTableBodyNone">No   </td><td class="markdownTableBodyNone">0.8   </td><td class="markdownTableBodyNone"><code>--threshold=0.7</code>    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone"><code>--norm=&lt;value&gt;</code>   </td><td class="markdownTableBodyNone">Normalization option for NEFF <br  />
 <b>0</b>: Normalize by the square root of sequence length <br  />
 <b>1</b>: Normalize by the sequence length <br  />
 <b>2</b>: No Normalization   </td><td class="markdownTableBodyNone">No   </td><td class="markdownTableBodyNone">0   </td><td class="markdownTableBodyNone"><code>--norm=2</code>    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyNone"><code>--omit_query_gaps=[true/false]</code>   </td><td class="markdownTableBodyNone">Omit gap positions of query sequence from entire sequences for NEFF computation   </td><td class="markdownTableBodyNone">No   </td><td class="markdownTableBodyNone">true   </td><td class="markdownTableBodyNone"><code>--omit_query_gaps=true</code>    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone"><code>--is_symmetric =true/false]</code>   </td><td class="markdownTableBodyNone">Consider gaps in number of differences when computing sequence similarity cutoff (asymmetric) or not (symmetric)   </td><td class="markdownTableBodyNone">No   </td><td class="markdownTableBodyNone">true   </td><td class="markdownTableBodyNone"><code>--is_symmetric=false</code>    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyNone"><code>--non_standard_option=&lt;value&gt;</code>   </td><td class="markdownTableBodyNone">Options for handling non-standard letters of the specified alphabet <br  />
 <b>0</b>: Treat them the same as standard letters <br  />
 <b>1</b>: Consider them as gaps when computing similarity cutoff of sequences (only used in asymmetryc version) <br  />
 <b>2</b>: Consider them as gaps in computing similarity cutoff and checking position of match/mismatch   </td><td class="markdownTableBodyNone">No   </td><td class="markdownTableBodyNone">0   </td><td class="markdownTableBodyNone"><code>--non_standard_option=1</code>    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone"><code>--depth=&lt;value&gt;</code>   </td><td class="markdownTableBodyNone">Depth of MSA to be considered in computation (starting from the first sequence)   </td><td class="markdownTableBodyNone">No   </td><td class="markdownTableBodyNone">inf (consider all sequences)   </td><td class="markdownTableBodyNone"><code>--depth=10</code> <br  />
(if given value is greater than original depth, it considers the original depth)    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyNone"><code>--gap_cutoff=&lt;value&gt;</code>   </td><td class="markdownTableBodyNone">Threshold for considering a position as gappy and removing that (between 0 and 1)   </td><td class="markdownTableBodyNone">No   </td><td class="markdownTableBodyNone">1 (no gappy position)   </td><td class="markdownTableBodyNone"><code>--gap_cutoff=0.7</code>    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone"><code>--pos_start=&lt;value&gt;</code>   </td><td class="markdownTableBodyNone">Start position of each sequence to be considered in NEFF (inclusive)   </td><td class="markdownTableBodyNone">No   </td><td class="markdownTableBodyNone">1 (the first position)   </td><td class="markdownTableBodyNone"><code>--pos_start=10</code>    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyNone"><code>--pos_end=&lt;value&gt;</code>   </td><td class="markdownTableBodyNone">Last position of each sequence to be considered in NEFF (inclusive)   </td><td class="markdownTableBodyNone">No   </td><td class="markdownTableBodyNone">inf (consider all sequence)   </td><td class="markdownTableBodyNone"><code>--pos_end=50</code> (if given value is greater than the length of the MSA sequences, consider length of sequences in the MSA)    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone"><code>--only_weights=[true/false]</code>   </td><td class="markdownTableBodyNone">Return only sequence weights, rather than the final NEFF   </td><td class="markdownTableBodyNone">No   </td><td class="markdownTableBodyNone">false   </td><td class="markdownTableBodyNone"><code>--only_weights=true</code>    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyNone"><code>--mask_enabled=[true/false]</code>   </td><td class="markdownTableBodyNone">Enable random sequence masking for NEFF calculation and return the masking with the highest NEFF   </td><td class="markdownTableBodyNone">No   </td><td class="markdownTableBodyNone">false   </td><td class="markdownTableBodyNone"><code>--mask_enabled=true</code>    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone"><code>--mask_percent=&lt;value&gt;</code>   </td><td class="markdownTableBodyNone">Percentage of sequences to be masked in each masking iteration   </td><td class="markdownTableBodyNone">when <em>mask_enabled</em>=true   </td><td class="markdownTableBodyNone">0   </td><td class="markdownTableBodyNone"><code>--mask_percent=0.4</code>    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyNone"><code>--mask_count=&lt;value&gt;</code>   </td><td class="markdownTableBodyNone">Number of masking iterations   </td><td class="markdownTableBodyNone">when <em>mask_enabled</em>=true   </td><td class="markdownTableBodyNone">0   </td><td class="markdownTableBodyNone"><code>--mask_count=0</code>    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone"><code>--multimer_MSA=[true/false]</code>   </td><td class="markdownTableBodyNone">Compute NEFF for paired MSA and individual monomer MSAs when MSA is in the form of multimer MSA   </td><td class="markdownTableBodyNone">No   </td><td class="markdownTableBodyNone">false   </td><td class="markdownTableBodyNone"><code>--multimer_MSA=true</code>    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyNone"><code>--monomer_length=&lt;list of values&gt;</code>   </td><td class="markdownTableBodyNone">Length of the monomers, used to obtain NEFF for paired MSA and individual monomer MSAs   </td><td class="markdownTableBodyNone">when <em>multimer_MSA</em>=true   </td><td class="markdownTableBodyNone">0   </td><td class="markdownTableBodyNone"><code>--monomer_length=17 45</code>    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone"><code>--column_neff=[true/false]</code>   </td><td class="markdownTableBodyNone">Compute Column-wise NEFF   </td><td class="markdownTableBodyNone">No   </td><td class="markdownTableBodyNone">false   </td><td class="markdownTableBodyNone"><code>--column_neff=true</code>   </td></tr>
</table>
<p>For more details about features, please refer to the <a href="https://maryam-haghani.github.io/Neffy/index.html#overview_neff_computation">documentation</a>.</p>
<h3><a class="anchor" id="autotoc_md8"></a>
Example:</h3>
<p>neff &ndash;file=alignment.fasta &ndash;threshold=0.6 &ndash;norm=2 &ndash;is_symmetric=false &ndash;check_validation=true As output, it will print the final MSA length, depth and Neff to the console, based on the given options.</p>
<p>For detailed instructions on usage, please refer to the <a href="https://maryam-haghani.github.io/Neffy/usage_guide.html#neff_computation">documentation</a>.</p>
<h2><a class="anchor" id="autotoc_md9"></a>
2. MSA File Conversion</h2>
<p>The MSA file conversion allows you to convert MSA files between different supported formats. <br  />
 All you need is to use the <code>converter</code> program and specify the input and output files with their formats, and the tool will perform the conversion accordingly.</p>
<h3><a class="anchor" id="autotoc_md10"></a>
Flags:</h3>
<p>The code accepts the following command-line flags: </p><table class="markdownTable">
<tr class="markdownTableHead">
<th class="markdownTableHeadNone">Flag   </th><th class="markdownTableHeadNone">Description   </th><th class="markdownTableHeadNone">Required   </th><th class="markdownTableHeadNone">Default Value   </th><th class="markdownTableHeadNone">Example    </th></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone"><code>--in_file=&lt;filename&gt;</code>   </td><td class="markdownTableBodyNone">Specifies the input MSA file to be converted.<br  />
 Replace <code>&lt;filename&gt;</code> with the path and name of the input file   </td><td class="markdownTableBodyNone">Yes   </td><td class="markdownTableBodyNone">N/A   </td><td class="markdownTableBodyNone"><code>--in_file=input.fasta</code>    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyNone"><code>--out_file=&lt;filename&gt;</code>   </td><td class="markdownTableBodyNone">Specifies the output file where the converted MSA will be saved.<br  />
 Replace <code>&lt;filename&gt;</code> with the desired path and name of the output file   </td><td class="markdownTableBodyNone">Yes   </td><td class="markdownTableBodyNone">N/A   </td><td class="markdownTableBodyNone"><code>--out_file=output.a2m</code>    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone"><code>--alphabet=&lt;value&gt;</code>   </td><td class="markdownTableBodyNone">Alphabet of MSA <br  />
 <b>0</b>: Protein <br  />
 <b>1</b>: RNA <br  />
 <b>2</b>: DNA   </td><td class="markdownTableBodyNone">No   </td><td class="markdownTableBodyNone">0   </td><td class="markdownTableBodyNone"><code>--alphabet=1</code>    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyNone"><code>--check_validation=[true/false]</code>   </td><td class="markdownTableBodyNone">Validate the input MSA file based on alphabet or not   </td><td class="markdownTableBodyNone">No   </td><td class="markdownTableBodyNone">true   </td><td class="markdownTableBodyNone"><code>--check_validation=true</code>   </td></tr>
</table>
<p>Please note that the conversion is performed based on the specified input and output file extensions. <br  />
 For more details about features, please refer to the <a href="https://maryam-haghani.github.io/Neffy/index.html#overview_converter">documentation</a>.</p>
<h3><a class="anchor" id="autotoc_md11"></a>
Example:</h3>
<p>Suppose you have an MSA file named "input.fasta" and you want to convert to the A2M format and save it as "output.a2m". </p><pre class="fragment">converter --in_file=input.fasta --out_file=output.a2m
</pre><p> For detailed instructions on usage, please refer to the <a href="https://maryam-haghani.github.io/Neffy/usage_guide.html#converter">documentation</a>.</p>
<h2><a class="anchor" id="autotoc_md12"></a>
Supported File Formats</h2>
<ul>
<li><b>A2M</b> (aligned FASTA-like format)</li>
<li><b>A3M</b> (compressed aligned FASTA-like format with lowercase letters for insertions)</li>
<li><b>FASTA</b>, <b>AFA</b>, <b>FAS</b>, <b>FST</b>, <b>FSA</b> (FASTA format)</li>
<li><b>STO</b> (Stockholm format)</li>
<li><b>CLUSTAL</b> (CLUSTAL format)</li>
<li><b>ALN</b> (ALN format)</li>
<li><b>PFAM</b> (format mostly used for nucleotides)</li>
</ul>
<p>In the <a href="https://maryam-haghani.github.io/Neffy/msa_formats.html">documentation</a>, you will find a brief explanation of each format, along with an illustrative alignment example for each one.</p>
<h2><a class="anchor" id="autotoc_md13"></a>
Error Handling</h2>
<p>If any errors occur during the execution of the MSA Processor, an error message will be displayed, describing the issue encountered. <br  />
 Please refer to the error message for troubleshooting or make necessary corrections to the input.</p>
<h1><a class="anchor" id="autotoc_md14"></a>
License</h1>
<p>This project is licensed under the GNU General Public License v3.0 - see the <a href="LICENSE.txt">LICENSE</a> file for details.</p>
<hr  />
<p> For further assistance, please see the <a href="https://maryam-haghani.github.io/Neffy/">documentation</a>. </p>
</div></div><!-- contents -->
</div><!-- PageDoc -->
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <title>Footer</title>
  <style>
    .footer {
      padding: 10px;
      font-size: 14px;
      color: #888;
    }
    .footer-left {
      text-align: left;
    }
    .footer address, .footer p {
      margin: 0;
    }
  </style>
</head>
<body>
  <div class="footer">
    <div class="footer-left">
      <hr/>
      <address>
        <small>
          Generated by&#160;<a href="https://www.doxygen.org/index.html">
          <img src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.11.0
        </small>
      </address>
      <p>© Copyright 2024, Maryam Haghani</p>
    </div>
  </div>
</body>
</html>
